### PR TITLE
tools: add fpgainfo support d5005 and N3000 pac card

### PR DESCRIPTION
**Pull request by anandaravuri**

### CODE REVIEW

File: tools/libboard/board_dc/board_dc.c [hunk 1]
```diff
@@ -42,13 +42,26 @@
 #include "board_dc.h"
 
 #define MACADDR_LEN 17
+#define FPGA_STR_SIZE     256
```
Comment (tswhison) [inline tools/libboard/board_dc/board_dc.c:45]: Could we reuse SYSFS_PATH_MAX?

File: tools/libboard/board_dc/board_dc.c [hunk 2]
```diff
@@ -129,7 +142,8 @@ fpga_result parse_fw_ver(char *buf, char *fw_ver, size_t len)
 
 	errno = 0;
 	var = strtoul(buf, &endptr, 16);
-	if (!endptr || *endptr != '\0') {
+	if (var == 0 &&
```
Comment (tswhison) [inline tools/libboard/board_dc/board_dc.c:145]: This should be..
`if (endptr != buf + strlen(buf)) {`

File: tools/libboard/board_dc/board_dc.c [hunk 3]
```diff
@@ -326,3 +340,126 @@ fpga_result print_board_info(fpga_token token)
 
 	return resval;
 }
+
+// Read sysfs
+fpga_result read_sysfs(fpga_token token, char *sysfs_path, char *sysfs_name)
```
Comment (tswhison) [inline tools/libboard/board_dc/board_dc.c:345]: When passing strings around, we should also pass the buffer size as a size_t.  Otherwise, we cannot know how large the storage behind these buffers actually is.

File: tools/libboard/board_dc/board_dc.c [hunk 4]
```diff
@@ -326,3 +340,126 @@ fpga_result print_board_info(fpga_token token)
...
+	}
+
+	res = fpgaObjectRead(sec_object, (uint8_t *)(&name), 0, size, 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to Read object ");
+		resval = res;
+		goto out_destroy;
+	}
+
+	snprintf(sysfs_name, FPGA_STR_SIZE, "%s", (char *)name);
```
Comment (tswhison) [inline tools/libboard/board_dc/board_dc.c:385]: This is using snprintf( ) to perform a string copy.  It should probably use strnlen( ), memcpy( ), and an explicit null termination instead.

File: tools/libboard/board_dc/board_dc.c [hunk 5]
```diff
@@ -326,3 +340,126 @@ fpga_result print_board_info(fpga_token token)
...
+		OPAE_MSG("Failed to get token Object");
+		return res;
+	}
+	printf("********** SEC Info START ************ \n");
+
+	// BMC Keys
+	memset(name, 0, sizeof(name));
+	if (read_sysfs(token, DFL_SYSFS_SEC_BMC_ROOT, name) == FPGA_OK)
+		printf("BMC root entry hash: %s", name);
+	else
```
Comment (tswhison) [inline tools/libboard/board_dc/board_dc.c:415]: This function returns an fpga_result.  Shouldn't these failed calls to read_sysfs affect the return value?

File: tools/libboard/board_dc/board_dc.h [hunk 6]
```diff
@@ -97,6 +97,29 @@ fpga_result parse_fw_ver(char *buf, char *fw_ver, size_t len);
...
+*
+* @param[in] token           fpga_token object for device (FPGA_DEVICE type)
+* @param[in] bmcfw_var       pointer to sysfs path
+* @param[inout] len          returns sysfs value as string
+
+* @returns FPGA_OK on success. FPGA_NOT_FOUND if invalid MAC address.
+* FPGA_INVALID_PARAM if invalid parameters were provided
+*
+*/
+fpga_result read_sysfs(fpga_token token, char *sysfs_path, char *sysfs_name);
```
Comment (tswhison) [inline tools/libboard/board_dc/board_dc.h:120]: Should this go into a generic library that can share the function across each of your libboard_X implementations?

File: tools/libboard/board_dc/board_dc.c (hunk 1 above)
Comment (anandaravuri) [reply tools/libboard/board_dc/board_dc.c]: fixed

File: tools/libboard/board_dc/board_dc.c (hunk 3 above)
Comment (anandaravuri) [reply tools/libboard/board_dc/board_dc.c]: moved common code libborad_common.so file

File: tools/libboard/board_dc/board_dc.c (hunk 4 above)
Comment (anandaravuri) [reply tools/libboard/board_dc/board_dc.c]: fixed

File: tools/libboard/board_dc/board_dc.c (hunk 3 above)
Comment (anandaravuri) [reply tools/libboard/board_dc/board_dc.c]: fixed

File: tools/libboard/board_dc/board_dc.h (hunk 6 above)
Comment (anandaravuri) [reply tools/libboard/board_dc/board_dc.h]: moved common code libborad_common.so file

File: tools/libboard/board_dc/board_dc.c (hunk 5 above)
Comment (anandaravuri) [reply tools/libboard/board_dc/board_dc.c:299]: fpgainfo out prints shows lot of error messages, so ignore errors.

File: tools/libboard/board_dc/board_dc.c (hunk 2 above)
Comment (anandaravuri) [reply tools/libboard/board_dc/board_dc.c:113]: if (endptr != buf + strlen(buf))  doesn't fix strtoul(), shows error message

File: tools/libboard/board_common/board_common.c [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+	}
+
+	res = fpgaObjectRead(fpga_object, (uint8_t *)(&name), 0, size, 0);
+	if (res != FPGA_OK) {
+		OPAE_ERR("Failed to Read object ");
+		resval = res;
+		goto out_destroy;
+	}
+
+	len = strnlen((char *)name, sizeof(name) - 1);
```
Comment (tswhison) [inline tools/libboard/board_common/board_common.c:86]:
```c
len = strnlen(name, len - 1);
memcpy(sysfs_name, name, len);
sysfs_name[len] = '\0';
```

File: tools/libboard/board_dc/board_dc.c (hunk 2 above)
Comment (tswhison) [reply tools/libboard/board_dc/board_dc.c:113]: What are the errors?

File: tools/libboard/board_dc/board_dc.c (hunk 5 above)
Comment (tswhison) [reply tools/libboard/board_dc/board_dc.c:299]: What is the process exit code when an error occurs?

File: tools/libboard/board_common/board_common.c (hunk 7 above)
Comment (anandaravuri) [reply tools/libboard/board_common/board_common.c]: fixed

File: tools/libboard/board_dc/board_dc.c (hunk 5 above)
Comment (anandaravuri) [reply tools/libboard/board_dc/board_dc.c]: fixed , exits with error code

File: tools/libboard/board_dc/board_dc.c (hunk 2 above)
Comment (anandaravuri) [reply tools/libboard/board_dc/board_dc.c]: fixed, driver ends with new line char (\n).

File: tools/libboard/board_common/board_common.h [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+#ifndef __FPGA_BOARD_COMMON_H__
+#define __FPGA_BOARD_COMMON_H__
+
+#include <opae/types.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+#define SYSFS_PATH_MAX 256
```
Comment (tswhison) [inline tools/libboard/board_common/board_common.h:36]: Why are you redefining this here?

File: tools/extra/fpgabist/bist/bist_dc.c [hunk 9]
```diff
@@ -265,7 +265,7 @@ int main(int argc, char *argv[])
 	fpga_token	   accelerator_token;
 	fpga_handle	   accelerator_handle;
 	int open_flags = 0;
-
+	uint32_t i = 0;
```
Comment (tswhison) [inline tools/extra/fpgabist/bist/bist_dc.c:268]: This file should go into a separate pull request.

File: tools/libboard/board_common/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+## SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
+## INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
+## CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
+## ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
+## POSSIBILITY OF SUCH DAMAGE.
+
+opae_add_static_library(TARGET board_common
+    SOURCE board_common.c
+    LIBS opae-c
+    COMPONENT opaeclib
```
Comment (tswhison) [inline tools/libboard/board_common/CMakeLists.txt:30]: Should probably be more diligent about using the correct COMPONENT names.

File: tools/libboard/board_common/board_common.h (hunk 8 above)
Comment (anandaravuri) [reply tools/libboard/board_common/board_common.h]: fixed
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif // HAVE_CONFIG_H

File: tools/extra/fpgabist/bist/bist_dc.c (hunk 9 above)
Comment (anandaravuri) [reply tools/extra/fpgabist/bist/bist_dc.c]: removed from merge request

File: tools/libboard/board_common/CMakeLists.txt (hunk 10 above)
Comment (anandaravuri) [reply tools/libboard/board_common/CMakeLists.txt]: Renamed to opaeboardlib

**Comment by tswhison** [comment]
Is there anything we can do to keep the code coverage from decreasing?

**Comment by anandaravuri** [comment]
> Is there anything we can do to keep the code coverage from decreasing?

[Ananda] Adding test cases to improve code coverage in another merge request.

File: tests/board/test_board_dc.cpp [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+#include <linux/ioctl.h>
+
+#include <cstdlib>
+#include <string>
+#include "gtest/gtest.h"
+#include "mock/test_system.h"
+#include "libboard/board_dc/board_dc.h"
+#include "libboard/board_common/board_common.h"
+
+#define SYSFS_MAX_SIZE         256
```
Comment (tswhison) [inline tests/board/test_board_dc.cpp:49]: Could you please use SYSFS_PATH_MAX by adding the following to your test file:
```c
#ifdef HAVE_CONFIG_H
#include <config.h>
#endif // HAVE_CONFIG_H
```

Comment (anandaravuri) [reply tests/board/test_board_dc.cpp]: fixed
